src/common/task_system/task_scheduler.cpp

```diff
@@ -144,5 +151,15 @@ void TaskScheduler::runWorkerThread() {
     }
 }
 
+void TaskScheduler::interruptTaskIfTimeOutNoLock(
+    processor::ExecutionContext* context, uint64_t scheduledTaskID) {
+    if (context->clientContext->isQueryTimeOut()) {
+        context->clientContext->interrupt();
+        removeErroringTask(scheduledTaskID);
+        throw common::InterruptException{common::StringUtils::string_format(
```

Comment:
Do not directly throw an exception here. It's better to `setException` for task. You can return a boolean flag in this function, then in the while loop
```
    while (!task->isCompleted()) {
        interruptTaskIfTimeOutNoLock(context, scheduledTask->ID);
        std::this_thread::sleep_for(
            std::chrono::microseconds(THREAD_SLEEP_TIME_WHEN_WAITING_IN_MICROS));
    }
```
whenever there is an interruption, break from the while loop. In this way, you can reuse current exception handling mechanism.
```
    if (task->hasException()) {
        removeErroringTask(scheduledTask->ID);
        std::rethrow_exception(task->getExceptionPtr());
    }
```

Reply:
As stated here: https://github.com/kuzudb/kuzu/pull/1395#discussion_r1144192680
We actually do not need to setException when timeout, the worker thread will set the exception when sees interrupt.